### PR TITLE
[CALCITE-5911] Add REGEXP_EXTRACT_ALL function (enabled in BigQuery library)

### PR DESCRIPTION
**Pull request by Anthrino**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java
Comment (tanclary) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:218]: Is this PR for REGEXP_EXTRACT and REGEXP_EXTRACT_ALL? Can you update the title/commit message if so

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (tanclary) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:438]: REGEXP_EXTRACT and REGEXP_EXTRACT_ALL are pretty similar. What do you think of anytime REGEXP_EXTRACT gets called, you can call REGEXP_EXTRACT_ALL method and then just retrieve the requested element (e.g. if the occurrence is 2 then retrieve the second element). Would this simplify things?

File: site/_docs/reference.md
Comment (tanclary) [inline site/_docs/reference.md:2783]: Surround with *

**Review by tanclary — COMMENTED** [review]
Left a few initial comments

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (Anthrino) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:438]: That makes sense @tanclary, I considered that approach but REGEXP_EXTRACT_ALL does not have the position or occurrence params like REGEXP_EXTRACT, so merging the functionality as you suggested would mean we find all occurrences of the pattern for every call to EXTRACT and then picking the required index (also these are optional params so many a times we would be only looking for the default first occurrence). This can be avoided by passing the occurrence to EXTRACT_ALL and stopping the loop on reaching that count, but we'd have to include some logic to skip this when we need the whole list of matches (extract_all). I felt this might be added complexity with some hit on the performance, hence decoupled it even though they have similar functionality. What do you think is it better to consolidate them?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java
Comment (Anthrino) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:218]: I know that's a bit confusing, this PR includes changes from EXTRACT cherrypicked here due to dependencies. Should be resolved when we merge that PR and rebase this one.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (tanclary) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:438]: That makes sense. I just noticed some of the logic is repeated between the functions so maybe at the least we can extract that into a helper or something. We should try to avoid repeated checks/logic across these methods if we can.

Comment (Anthrino) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:418]: Hi @tanclary, moved the redundant error checks into a couple of helper functions to reuse majority of the code, I've kept the matching logic separate due to the above reasons would be good if you can have a look and suggest any improvements if needed.

Comment (tanclary) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:418]: That sounds good, once we get REGEXP_EXTRACT merged and can rebase this I'll come back for a second review 👍

Comment (Anthrino) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: @tanclary I've rebased the changes from previous PRs, also consolidated few of the redundant error checks into helper methods would be good if you can give it a review!

**Review by tanclary — COMMENTED** [review]
I think this looks pretty good..no additional comments at first glance but I will wait a bit to see if anyone else has comments before I merge.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (ILuffZhe) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:4692]: In your tests, I noticed that either the value or the regex is null, the function returns null.
What if we use `ReturnTypes.ARG0_NULLABLE` for this func?

**Review by ILuffZhe — COMMENTED** [review]
Overall looks good, just one minor comment.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (Anthrino) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:4692]: Good catch @ILuffZhe, thanks! 
Made the change in sql lib operator definition, I missed that assuming its always returning an array (empty for no matches).

**Review by tanclary — APPROVED** [review]
Looks good, Jerin. Thanks for your patience on this.

**Pull request merged**
